File: leetcode/easy/155_min_stack.py
# -----------------------------------------
# My Solution
#
# Time  Complexity: O(n) for pop, O(1) for others
# Space Complexity: O(n)
# -----------------------------------------
class MinStack:

    def __init__(self):
        self.stack   = []
        self.min_val = float('inf')

    def push(self, x: int) -> None:
        self.stack.append(x)
        self.min_val = min(self.min_val, x)

    def pop(self) -> None:
        pop_val = self.stack.pop()
        if pop_val == self.min_val:
            self.min_val = min(self.stack) if len(self.stack) > 0 else float('inf')

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_val

# -----------------------------------------
# Model Solution
#
# Time  Complexity: O(1)
# Space Complexity: O(n)
# -----------------------------------------
class MinStack:

    def __init__(self):
        self.main_stack = []
        self.min_stack  = []

    def push(self, x: int) -> None:
        self.main_stack.append(x)
        if len(self.min_stack) == 0 or x <= self.min_stack[-1]:
            self.min_stack.append(x)

    def pop(self) -> None:
        pop_val = self.main_stack.pop()
        if pop_val == self.min_stack[-1]:
            self.min_stack.pop()

    def top(self) -> int:
        return self.main_stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]

File: leetcode/easy/155_min_stack.cpp
/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

// -----------------------------------------
// Model Solution
//
// Time  Complexity: O(1)
// Space Complexity: O(n)
// -----------------------------------------
// Ref: https://leetcode.com/problems/min-stack/discuss/49016/C%2B%2B-using-two-stacks-quite-short-and-easy-to-understand

class MinStack {
public:
    /** initialize your data structure here. */
    MinStack() {}
    
    void push(int val) {
        main_stack.push_back(val);
        if (min_stack.empty() || val <= min_stack.back())
            min_stack.push_back(val);
    }

    void pop() {
        int value_to_pop = main_stack.back();
        main_stack.pop_back();
        if (value_to_pop == min_stack.back())
            min_stack.pop_back();
    }

    int top() {
        return main_stack.back();
    }

    int getMin() {
        return min_stack.back();
    }

private:
    vector<int> main_stack;
    vector<int> min_stack;
};
